models/attachment.go

```diff
@@ -1,11 +1,149 @@
+func (a Attachment) ValidateAttachment() error {
+
+	ptx := PhishingTemplateContext{
+		BaseRecipient: BaseRecipient{
+			FirstName: "Foo",
+			LastName:  "Bar",
+			Email:     "foo@bar.com",
+		},
+		BaseURL:     "http://testurl.com",
+		URL:         "http://testurl.com/?rid=1234567",
+		TrackingURL: "http://testurl.local/track?rid=1234567",
+		Tracker:     "<img alt='' style='display: none' src='http://testurl.local/track?rid=1234567'/>",
+		From:        "Foobus Barbus",
+		RId:         "1234567",
+	}
+
+	_, err := a.ApplyTemplate(ptx)
+	return err
+}
+
+// ApplyTemplate parses different attachment files and applies the supplied phishing template.
+func (a *Attachment) ApplyTemplate(ptx PhishingTemplateContext) (io.Reader, error) {
+
+	var processedAttachment string
+	decodedAttachment, err := base64.StdEncoding.DecodeString(a.Content) // Decode the attachment
+	if err != nil {
+		return nil, err
+	}
+
+	// If we've already determined there are no template variables in this attachment return it immediately
+	if a.vanillaFile == true {
+		return strings.NewReader(string(decodedAttachment)), nil
+	} else {
+
+		// Decided to use the file extension rather than the content type, as there seems to be quite
+		//  a bit of variability with types. e.g sometimes a Word docx file would have:
+		//   "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
+		fileExtension := filepath.Ext(a.Name)
+
+		switch fileExtension {
+
+		case ".docx", ".docm", ".pptx", ".xlsx", ".xlsm":
+			// Most modern office formats are xml based and can be unarchived.
+			// .docm and .xlsm files are comprised of xml, and a binary blob for the macro code
+
+			// Create a new zip reader from the file
+			zipReader, err := zip.NewReader(bytes.NewReader(decodedAttachment), int64(len(decodedAttachment)))
+			if err != nil {
+				return nil, err
+			}
+
+			newZipArchive := new(bytes.Buffer)
+			zipWriter := zip.NewWriter(newZipArchive) // For writing the new archive
+
+			// i. Read each file from the Word document archive
+			// ii. Apply the template to it
+			// iii. Add the templated content to a new zip Word archive
+			fileContainedTemplatesVars := false
+			for _, zipFile := range zipReader.File {
+				ff, err := zipFile.Open()
+				if err != nil {
+					return nil, err
+				}
+				defer ff.Close()
+				contents, err := ioutil.ReadAll(ff)
+				if err != nil {
+					return nil, err
+				}
+				subFileExtension := filepath.Ext(zipFile.Name)
+				var tFile string
+				if subFileExtension == ".xml" || subFileExtension == ".rels" { // Ignore other files, e.g binary ones and images
+					// For each file apply the template.
+					tFile, err = ExecuteTemplate(string(contents), ptx)
+					if err != nil {
+						return nil, err
+					}
+					// Check if the subfile changed. We only need this to be set once to know in the future to check the 'parent' file
+					if tFile != string(contents) {
```

Comment:
If we only need to set this once, we should check if it's set in this condition, e.g.:

```go
if !fileContainedTemplateVars && tFile != string(contents) { ... }
```

Reply:
And out of curiosity, why don't we just set the (renamed) `a.isTemplate = true` here instead of having both `fileContainedTemplatesVars` and `a.isTemplate`?

Reply:
That's some satisfying short circuiting there!

Hmm, I think I created `fileContainedTemplatesVars` first before implementing your more succinct `a.isTemplate` approach, and didn't merge the ideas.

Reply:
I realise the problem with your suggestion. 

It boils down to needing three states for a file;

0 --> File has never been checked, so we don't know if it has template variables (= apply template logic)
1 --> File has template variables (= apply template logic)
2 --> File does not have template variables (= return file)

To get around this, and use a regular boolean, I assume that a file *does* have template variables (thus the name `isVanilla` as this defaults to false - indicating template variables *are* present. This ensures files we haven't checked before are parsed, and if no template variables are found we can flip this default boolean. We have to do this kind of reverse logic as (AFAIK) we can only set the default state of a boolean to false.

When we load up a file the first thing we do is check if we have previously ascertained that is does not contain template variables, and can safely return it as is:

```
if a.vanillaFile == true {
   return decodedAttachment, nil
} else {
  ....
```

As we parse over a zip file's contents we check each file in the archive to see if its content changed after applying the template, and is therefore a templated file. We are interested in the inverse, however, as our initial assumption is that it *is* a templated file. So, if the content has remained the same for *every* file in the zip archive, we can then finally set `a.vanillaFile` to `true`, away from its default of `false`. This is why we need that temporary boolean `fileContainedTemplateVars`.

Of course the alternative is to have `a.isTemplate` as an `int` with the three states mentioned above. I'm not sure which approach is more readable, but given the trickiness of explaining the above, the tristate might be easier.

The other other alternative is that I'm talking nonsense and there's a trivial solution I'm not seeing 😅